### ISSUE TITLE
Offer a dedicated sync() interface on the base Metric class

### ISSUE THREAD

**Issue by ananthsub** [PyTorchLightning/metrics#67]
## 🚀 Feature
Offer a dedicated `sync()` interface on the base Metric class. This would consolidate state across a provided process group using a given `dist_sync_fn` and would let us deprecate the `dist_sync_on_step` flag on the metric constructor.

### Motivation

The reason we'd like this is to decouple metric computation and global syncing. As a result, we'd be able to inspect both the local metric state separately from the synced state. 
Example scenario:
- We're training on a large number of nodes
- We wish to create a metric to track the local state during training steps, as syncing each step will be incredibly expensive
- At the end of the epoch, we want to sync the state once and log this value.

This interface also enables the training framework to offer higher-level APIs that could automatically call `sync()` for a particular `Metric` at relevant spots in the training loop (e.g. on_step, or on_epoch in Lightning).

cc @maximsch2 

### Pitch

We should be able to re-use most of `_sync_dist()` already. 

### Alternatives
Keep as is
